### ISSUE TITLE
Sync to Device queues episodes but does not complete transfers (gtk3 branch)

### ISSUE THREAD

**Issue by dannn-o** [gpodder/gpodder#279]
Running gPodder (ver. 3.9.3) as a git clone with the gtk3 branch (Linux Mint 17.3, libgtk-3 ver. 3.10.8), the "Sync to device" option is queuing up the new episodes to transfer, and the corresponding playlists are created successfully on the flash drive (as is specified in the Preferences), but gPodder then stops without actually transferring the files.  The gPodder window appears to be functional (i.e. responds to mouse clicks to navigate through menus, etc), but the Progress tab stays "stuck" with all episodes in a Queued status without the progress indicators advancing:
![sync hang](https://cloud.githubusercontent.com/assets/17674969/25207983/b6b5af3c-253f-11e7-8168-4259ec67d840.png)

If I try to exit gPodder, a warning is issued that episodes are being downloaded:
![screenshot from 2017-04-19 20 38 15](https://cloud.githubusercontent.com/assets/17674969/25208062/31cdbe6c-2540-11e7-8126-8326067821d8.png)

I went ahead and clicked Quit, which exited successfully.  Attached is the corresponding log file:

[2017-04-19.txt](https://github.com/gpodder/gpodder/files/939008/2017-04-19.txt)

**Comment by vossad01** [comment]
This will take some investigation, unfortunately this is no exception of other obvious failure.  I swore I had problems with sync before `gtk3` which is part of why I have not looked into this previously.

As a possible workaround, there you can select multiple and use "Send to > Local Folder" on the right-click context menu.  On the `gtk3` you will only be prompted for the folder to save all the files, instead of being asked where to save each individual file.  This does not help with the playlist though.

**Comment by vossad01** [comment]
In watching this in the debugger, it seems that [this loop](https://github.com/gpodder/gpodder/blob/917c33f72bad4574ba84cbfec42b4c3245dc9099/src/gpodder/gtkui/main.py#L1141) is never entered which at least points to why it will indicate it is done despite there being queued tasks.

**Comment by dannn-o** [comment]
I have to say:  sync always worked solidly for me with the original version I was running from the thp/gpodder PPA.  As a workaround, I've been manually copying the episodes from the gPodder/Downloads folder to the flash drive which works, but doesn't benefit from the awesome renaming normally provided by gPodder.

I gave your suggestion of using the "Send to" option in the right-click context menu, and discovered something which might help to isolate the problem:  when I invoked "Send to" with a destination of the flash drive, nothing at all happened after clicking "Save".  When I did the same to a different location (e.g. Desktop), it worked fine.  I'm again attaching the log file with the hope that it will help:
[2017-04-20.txt](https://github.com/gpodder/gpodder/files/942049/2017-04-20.txt)

I attempted the "Send to" with the same episode (Lore - Episode 57) which is why you'll see multiple entries.

**Comment by mxyzplx** [comment]
I have this problem ALL. THE. TIME.

One thing to try is to just "Sync to device" a second or sometimes a third time. gPodder won't necessarily show the progress of all or any of the files being transferred (I usually see a cue, three files transferring part-way, and nothing else, it just stops there), but it WILL finish transferring, even if the gui doesn't represent that accurately.

**Comment by dannn-o** [comment]
@mxyzplx -  Unfortunately, retrying the Sync does nothing in my case but queue up additional copies of the same episodes.  I understand you've had Sync problems before, but in my particular case, I believe this is something related to the GTK3 branch I recently switched to: either a bug, or an environmental / library conflict between my git clone and non-git clone environments.

@vossad01 - After considering it further, I'm not so sure that the "Send To" failure is necessarily related to the Sync problem.  If you think it merits a separate issue, let me know.  And, if there is anything further I can do or test to help narrow the scope of the Sync problem, I'd be glad to try whatever.

**Comment by vossad01** [comment]
@dannn-o You are right that the "Send To" failure is different.  I have been meaning to fix that for a little while now and have opened #282.  See that issue for the workaround I have been using.

@mxyzplx Thanks for reporting that @dannn-o's issue may (or may not) be different but I am reassured to know I am not the only one who has had problems with Sync.

**Comment by dannn-o** [comment]
Just to keep you updated:  I went ahead and switched back (using 'git checkout master') to the original Python 2 gPodder version I had been using, and the Sync to Device feature is now working again as it had been.  I'm a bit of a podcast junkie....had to get my "fix" loaded on the flash drive for my commute tomorrow  :0)

Of course, I'm back to dealing with the original problem of the SSLV3 alert handshake failure, but figured I could use Clementine for now to work around that.  I'd be glad to switch back to the git clone for purposes of testing if you make progress on the gtk3 fix for Sync.  And, thanks for the update on the 'Send To' issue...the workaround you suggested is exactly what I found worked for me.  Switching back to the older gPodder version to do the Sync in bulk was the path of least resistance for me, for now.

**Comment by vossad01** [comment]
@dannn-o While not convenient, I suppose you would have the option of switching between the branches: `master` for syncing and `gtk3` for updating.  I don't know what will happen with them during review, so they may not be in their final form, but there are now pull requests for the Sen to Folder (PR #283) and Sync (PR #287) issues that you could check out if you wanted a more immediate solution.

**Comment by elelay** [comment]
@dannn-o Adam's fix has landed in gtk3, so please test it when you have time.
Thanks,

**Comment by dannn-o** [comment]
OK - I switched back to the gtk3 branch (sudo git checkout gtk3), then ran `sudo git pull` to get up-to-date.  I then launched gPodder and ran through a Sync to Device.  The good news:  the Progress tab no longer hung up, and all of the pending episodes were processed.  The not-so-good news:  Not all of the episodes were actually transferred successfully.  It appears to be the ones that have 'special characters' (i.e. a colon, usually) as part of their name.   Those without any special characters transferred successfully. 

For the "problematic" episodes:  the .m3u file shows the episode(s) that were supposed to transfer, but the destination folder for these episodes is empty after the sync is finished.  Not sure if it will be helpful, but I'm attaching the resulting log file (2017-04-25.txt):
[2017-04-25.txt](https://github.com/gpodder/gpodder/files/956997/2017-04-25.txt)

Let me know if I can help with any further diagnostic data or testing.

**Comment by vossad01** [comment]
@dannn-o Thanks for checking testing this!  This seems overly pedantic, but I have opened #295 for the issue of those excluded episodes.  I suspected that issue existed when working on #282 but didn't bother to check.  I am not leaving this one open for that problem as well because I also suspect that problem exists outside of the `gtk3` branch and this issue is `gtk3` specific.

However, if you hadn't noticed, the issue with "Sent to Folder" and those 'special characters' has been fixed.

As an aside, have you placed your repository somewhere you don't have write access?  It is not typical to need `sudo` when executing Git commands.  If you know what you are doing and it makes sense for your use case thats fine, I just thought I would point it out in case it was incidental.

**Comment by mxyzplx** [comment]
For what it's worth, "sync to device" a second time on gpodder (gtk3 branch) will often cue up additional copies of the same episodes I already tried syncing. But only one of each episode gets synced, and the sync will finish.

Sometimes I have to do it a third time, but I have yet to have that fail.

**Comment by vossad01** [comment]
@mxyzplx Thank you for the testing and information.
I want to make sure I understand the issue.  I have confirmed that if clicking "Sync to device" a second time can result in the episode being queued for transfer a second time.  I am not sure what you mean by "Sometimes I have to do it a third time."  Are you saying queuing these duplicate transfers has a purpose/reason and sometimes you need them queued in triplicate?  I was thinking the bug was that duplicate entries could appear at all.

**Comment by dannn-o** [comment]
@vossad01 - I may be understanding some of the relationships between the different issues involved, but thought this might be worth mentioning:

You stated that you thought the issue now being addressed with #295 might exist outside of the gtk3 branch, but I do not encounter it when I revert back to the non-gtk3 (Master) branch.  As an added confirmation, I just switched back to Master, was notified that my branch was behind and could be fast-forwarded (which I did using "git pull"), and then I launched gPodder and invoked Sync to Device.  The Sync completed successfully, including those episodes with 'special' characters (again, mostly a colon).  If a colon existed, it was replaced by an underscore.  So, it seems to me that gPodder has logic to handle VFAT-friendly naming in the Master branch, and so the issue may actually be specific to the GTK3 branch, as this issue (#279) is (?)

Anyway, in response to my using "SUDO":  I did install the git clone into a protected directory (\usr\src), which I thought was a more managed location to help keep things structured and protected.  And again, I recognize what @mxyzplx is saying about multiple sync requests, but I am not encountering that issue.  With the fix for #279 applied in the gtk3 branch, and whenever I switch back to the Master branch, I only have to run the Sync to Device one time.
